docs/docs/learn-by-example/1-writing-an-initial-program/README.md

```diff
@@ -0,0 +1,107 @@
+
+Let's start by writing an early simple version of our program to start from there. Let's image a `UserDatabase` for the social network so new users can register:
+
+```kotlin
+interface UserDatabase {
+  fun createUser(name: String): UserId
+  fun findUser(userId: UserId): User?
+}
+```
+
+If our backend was a **Resful API**, we would expect it to offer an endpoint to create new users that would rely on this database.
+
+If we wanted to provide a somehow controlled implementation for this database contract that **did not rely on third parties**, one option could be to provide an in-memory implementation for it:
+
+```kotlin
+object InMemoryUserDatabase : UserDatabase {
+  private var users: List<User> = emptyList()
+
+  override fun createUser(name: String): UserId {
+    val userId = generateId(name)
+    this.users = users + listOf(User(userId, name))
+    return userId
+  }
+
+  override fun findUser(userId: UserId): User? =
+    users.find { it.id == userId }
+
+  private fun generateId(name: String): UserId = 
+    UserId("$name${UUID.randomUUID()}")
+}
+```
+
+This implementation for the `UserDatabase` contains a mutable state that is mutated by the `createUser` function. It provides functions to `createUser` and `findUser` in the database. The `createUser` function relies on a private utility function to generate a unique `UserId` from a given `name`.
+
+So we could have a program to rely on this database to create and find users. Feel free to click on the ▶️ icon to run it and check the result.
+
+```kotlin:ank:playground
+import java.util.*
+
+class UserId(val id: String)
+data class User(val id: UserId, val name: String)
+
+interface UserDatabase {
+  fun createUser(name: String): UserId
+  fun findUser(userId: UserId): User?
+}
+
+object InMemoryUserDatabase : UserDatabase {
+  private var users: List<User> = emptyList()
+
+  override fun createUser(name: String): UserId {
+    val userId = generateId(name)
+    this.users = users + listOf(User(userId, name))
+    return userId
+  }
+
+  override fun findUser(userId: UserId): User? =
+    users.find { it.id == userId }
+
+  private fun generateId(name: String): UserId = UserId("$name${UUID.randomUUID()}")
+}
+
+fun main() {
+  //sampleStart
+  val userId = InMemoryUserDatabase.createUser("SomeUserName")
+  println("UserId: $userId")
+  val user = InMemoryUserDatabase.findUser(userId)
+  println(user)
+  //sampleEnd
+}
+``` 
+
+As you can see, we could register and find a user in the database. But this program has a few issues.
+
+We are not handling any potential errors, because we assumed it's an in memory implementation that will not fail. But what If it was a real database? **It could potentially throw, making our function calls non deterministic**. Also, what happens when the user we are trying to find is not there? If our backend was a Resful API, we would probably want to handle it by returning a 404 not found response to the client. 
```

Comment:
perhaps introduce that the call could be async and the current implementation forces to block

Reply:
Right, added now 👍